docs/developer/best-practices/navigation.asciidoc

```diff
@@ -0,0 +1,201 @@
+        onClick={(e) => {
+            e.preventDefault();
+            core.application.navigateToUrl(`/dashboard/my-dashboard`); 
+        }}> 
+        Go to Dashboard 
+    </a>; 
+----
+
+
+It is too much boilerplate to do this for each {kib} link in your app, so there is a handy HOC that helps with it: 
+{kib-repo}tree/{branch}/src/plugins/kibana_react/public/app_links/redirect_app_link.tsx#L49[RedirectAppLinks].
+
+[source,typescript jsx]
+----
+const MyApp = () => 
+  <RedirectAppLinks application={core.application}>
+    {/*...*/}
+        <a href={`/dashboard/my-dashboard`}>Go to Dashboard</a> {/* navigations using this link will happen in SPA friendly way */}
+    {/*...*/}
+  </RedirectAppLinks>
+----
+
+
+[[deep-linking]]
+=== Deep-linking into {kib} apps
+
+**Consider a {kib} app URL a part of app's plugin contract.** +
+Try to avoid hardcoding other app's URL in your app's code:
+
+[source,typescript jsx]
+----
+const urlToDashboarWithComplexState = `/dashboards#/view/0289f5d0-e61c-11ea-8c31-8bd3114ed023?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-15m,to:now))&_a=(description:'',filters:!(),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),query:(language:kuery,query:''),timeRestore:!f,title:'URL%20Drilldown%20Demo',viewMode:view)`
+----
+
+Instead, each app should expose {kib-repo}tree/{branch}/src/plugins/share/public/url_generators/README.md[a URL generator].
+Other apps should use those URL generators for deep-linking.
+To get a better idea, take a look at {kib-repo}tree/{branch}/src/plugins/dashboard/public/url_generator.ts#L38[dashboard's app URL generator].
+It allows specifying various dashboard app state pieces like: dashboardId, filters, query, time range and more.
+And then others should use it to generate a link a dashboard with specified state:
+
+[source,typescript jsx]
+----
+const dashboardUrl = dashboardUrlGenerator.createUrl({dashboardId, filters, timeRange});
+----
+
+There are two ways to access other's app URL generator in your code:
+
+1. *(preferred)* From a plugin contract of a destination app.
+2. (in case an explicit plugin dependency is not possible) Using {kib-repo}tree/{branch}/src/plugins/share/public/url_generators/README.md[url generator service] from `share` plugin. 
+
+
+[[routing]]
+=== Setting up internal app routing
+
+It is very common for {kib} apps to use React and React Router.
+Common rules to follow in this scenario:
+
+* Set up `BrowserRouter` and not `HashRouter`;
+* *Initialize your router with `history` instance provided by the `core`.*
+
+This is required to make sure `core` is aware of navigations triggered inside your app, so it could act accordingly when needed.
+
+* `Core`'s {kib-repo}tree/{branch}/docs/development/core/public/kibana-plugin-core-public.scopedhistory.md[ScopedHistory] instance
+* {kib-repo}tree/{branch}/test/plugin_functional/plugins/core_plugin_a/public/application.tsx#L120[Example plugin]
```

Comment:
Not sure, but may want to add link to `docs/development/core/public/kibana-plugin-core-public.appmountparameters.history.md` and / or the source, as it also got some usage examples.

https://github.com/elastic/kibana/blob/e976bddf81d5c811daa47f703fb474d3c601bec3/src/core/public/application/types.ts#L367-L376